update ship API, add getBoxesById()

JoeArmani opened the pull request:
Related: https://github.com/episphere/connect/issues/625
-The ship api has been failing intermittently with no error response. Boxes are not updated as shipped, so the site has to force the shipment with a fake tracking number. This is not ideal or scalable. 

-inspected and updated `ship` endpoint with notes.
-updated `shipBatchBoxes()` function to ensure shipped property is set and run as a Firestore transaction (atomic, ensures data integrity, automatically retries on initial failure).
-added `getBoxesById()` to support get requests on boxes with or without transaction. This will be useful in other scenarios as well.

Note: rules for firestore `'in'` query: max 30 disjunctions. This is preferred because it's sent as a single query. If over 30 disjunctions, we process the request with `Promise.all()`, which is sent as `n` queries.
Disjunctions = (items in array) * (where() clauses).

ramavavi commented:
Hi Joe, In the code, array of boxIdAndShipmentDataArray into a boxIdToShipmentData object. What is the reason for this change?

JoeArmani commented:
> Hi Joe, In the code, array of boxIdAndShipmentDataArray into a boxIdToShipmentData object. What is the reason for this change?

Hi @ramavavi, in the endpoint, I was trying to change as little as possible, so I just changed that array from let to const.

In the shipBatchBoxes() function, we updated it to a lookup table by boxId and shipment data for faster lookup and merging during the get/write transaction.

Gbarra9 commented on utils/firestore.js:
```diff
@@ -1309,35 +1309,83 @@ const updateTempCheckDate = async (institute) => {
+                if (shipmentData) {
+                    shipmentData[fieldMapping.submitShipmentFlag] = fieldMapping.yes;
+                    Object.assign(boxData, shipmentData);
+                    transaction.update(box.docRef, boxData);
```
Initially, I was curious why you were using map to transform each document to `{ data: document.data(), docRef: document.ref }` in the `getBoxesByBoxId` function control flow. This line of code explains why you needed that document.ref, it's for updating. Also, I never used or seen `db.runTransaction`. Thanks for the jsdocs and using it so I can read up on it.

JoeArmani replied:
Thanks, Gene. Normally, I think the previous batch write was substantial enough for this operation. Since this has been a persistent issue, I like the idea of using a transaction here and tightening up the entire call chain end-to-end. It's a little slower, but we're likely talking about milliseconds. I like that transactions auto-retry and lock access to the associated docs while the transaction runs...it's just an extra layer of certainty that our shipped box data will be consistent.

we-ai replied:
The problem is that, based on my checks on thousands lines of logs, there were no backend errors in those failed shipments: `ship` API was not called. I believe the problem lies in frontend code.

Gbarra9 reviewed (approved):
Looks good. No further suggestions.

we-ai commented on utils/biospecimen.js:
```diff
@@ -359,70 +360,89 @@ const biospecimenAPIs = async (req, res) => {
         };
 
-        let boxIdAndShipmentDataArray = [];
+        const boxIdAndShipmentDataArray = [];
```
I prefer `let` here, to indicate the value may change.

we-ai commented on utils/biospecimen.js:
```diff
@@ -359,70 +360,89 @@ const biospecimenAPIs = async (req, res) => {
-            ...sharedShipmentData,
-            959708259: boxIdToTrackingNumberMap[boxId],
-          };
+            const shipmentData = {
```
I prefer `let` here, to indicate the value may change.

JoeArmani replied:
Updated to Warren's preference.

we-ai commented on utils/biospecimen.js:
```diff
@@ -359,70 +360,89 @@ const biospecimenAPIs = async (req, res) => {
-          return res.status(500).json({ message: 'Internal Server Error', code: 500 });
-        }
-        
+            console.error(`Error occurred when running shipBatchBoxes(): ${error}`);
```
```suggestion
            console.error(`Error occurred when running shipBatchBoxes(). ${error}`);
```

JoeArmani replied:
Updated to Warren's preference.

we-ai commented on utils/firestore.js:
```diff
@@ -1309,35 +1309,83 @@ const updateTempCheckDate = async (institute) => {
+            const snapshot = await getSnapshot(db.collection('boxes').where(shippingBoxId, 'in', boxIdArray).where(loginSite, '==', siteCode));
+            return snapshot.docs.map(document => ({ data: document.data(), docRef: document.ref }));
+        } else {
+            const queries = boxIdArray.map(boxId => {
+                return db.collection('boxes').where(shippingBoxId, '==', boxId).where(loginSite, '==', siteCode);
+            });
+        
+            const boxResults = await Promise.all(queries.map(query => getSnapshot(query)));
+            return boxResults.map(result => result.docs.map(document => ({ data: document.data(), docRef: document.ref }))).flat();
```
The 'else' block handles mixed scenarios: BPTL ( when size>30) and non BPTL (when size>15). Please make sure `siteCode` is there for BPTL scenario calls.

JoeArmani replied:
Update: Warren's comment makes me remember this function is useless for BPTL's needs because `boxId` is not unique across sites. I was building this for later use, but it will not be useful. Removed `isBPTL` entirely.

we-ai commented on utils/firestore.js:
```diff
@@ -1309,35 +1309,83 @@ const updateTempCheckDate = async (institute) => {
 
-    if (snapshot.size !== 1) {
-      return false;
+        if (boxIdArray.length <= 30 && isBPTL) {
+            const snapshot = await getSnapshot(db.collection('boxes').where(shippingBoxId, 'in', boxIdArray));
+            return snapshot.docs.map(document => ({ data: document.data(), docRef: document.ref }));
+        } else if (boxIdArray.length <= 15 && !isBPTL) {
+            const snapshot = await getSnapshot(db.collection('boxes').where(shippingBoxId, 'in', boxIdArray).where(loginSite, '==', siteCode));
+            return snapshot.docs.map(document => ({ data: document.data(), docRef: document.ref }));
+        } else {
+            const queries = boxIdArray.map(boxId => {
+                return db.collection('boxes').where(shippingBoxId, '==', boxId).where(loginSite, '==', siteCode);
+            });
+        
+            const boxResults = await Promise.all(queries.map(query => getSnapshot(query)));
+            return boxResults.map(result => result.docs.map(document => ({ data: document.data(), docRef: document.ref }))).flat();
```
We better have a consistent way to handle different size of data. Same comment here as [there](https://github.com/episphere/connectFaas/pull/427#discussion_r1334779671).

JoeArmani replied:
Update: Warren's suggested pattern is better. Updated to chunking when `boxIdArray.length > 15` and using `Promise.all()` to handle chunked `get()` requests.

we-ai commented on utils/firestore.js:
```diff
@@ -1309,35 +1309,83 @@ const updateTempCheckDate = async (institute) => {
+ * 
+ * @param {Array<string>} boxIdArray - array of box ids to fetch 
+ * @param {string} siteCode - site code of the user (number)
+ * @param {*} isBPTL - boolean to indicate if the request is from BPTL
```
```suggestion
 * @param {boolean} isBPTL - boolean to indicate if the request is from BPTL
```

JoeArmani replied:
Updated doc (removed isBPTL entirely).

we-ai reviewed (approved):
Looks good. Comments on minor things.
